File: RentalModule.Infrastructure/Mappers/CarOfferPersistenceMapper.cs
using RentalModule.Domain.Models;
using RentalModule.Infrastructure.DataBase.Entities;
using Shared.Utilities;

namespace RentalModule.Infrastructure.Mappers
{
    public class CarOfferPersistenceMapper : IPersistenceMapper<CarOfferEntity, CarOfferModel>
    {
        private readonly IPersistenceMapper<CarTagEntity, CarTagModel> _tagMapper;
        private readonly IPersistenceMapper<ImageUrlEntity, ImageUrlModel> _imageUrlMapper;
        private readonly IPersistenceMapper<TimePeriodEntity, TimePeriodModel> _timePeriodMapper;
        private readonly IPersistenceMapper<CarTariffEntity, CarTariffModel> _tariffMapper;

        public CarOfferPersistenceMapper(
            IPersistenceMapper<CarTagEntity, CarTagModel> tagMapper,
            IPersistenceMapper<ImageUrlEntity, ImageUrlModel> imageUrlMapper,
            IPersistenceMapper<TimePeriodEntity, TimePeriodModel> timePeriodMapper,
            IPersistenceMapper<CarTariffEntity, CarTariffModel> tariffMapper)
        {
            _tagMapper = tagMapper;
            _imageUrlMapper = imageUrlMapper;
            _timePeriodMapper = timePeriodMapper;
            _tariffMapper = tariffMapper;
        }

        public CarOfferModel MapToModel(CarOfferEntity entity) => new()
        {
            Id = entity.Id,
            Heading = entity.Heading,
            ShortDescription = entity.ShortDescription,
            FeaturedImageUrl = entity.FeaturedImageUrl,
            UrlHandle = entity.UrlHandle,
            Horsepower = entity.Horsepower,
            YearOfProduction = entity.YearOfProduction,
            EngineDetails = entity.EngineDetails,
            DriveDetails = entity.DriveDetails,
            GearboxDetails = entity.GearboxDetails,
            CarDeliverylocation = entity.CarDeliverylocation,
            CarReturnLocation = entity.CarReturnLocation,
            PublishedDate = entity.PublishedDate,
            Visible = entity.Visible,
            Tarrif = _tariffMapper.MapToModel(entity.Tarrif),
            Tags = entity.Tags?.Select(tag => _tagMapper.MapToModel(tag)).ToArray() ?? Array.Empty<CarTagModel>(),
            ImageUrls = entity.ImageUrls?.Select(image => _imageUrlMapper.MapToModel(image)).ToList() ?? new List<ImageUrlModel>(),
            UnavailableDates = entity.UnavailableDates?.Select(period => _timePeriodMapper.MapToModel(period)).ToList() ?? new List<TimePeriodModel>()
        };

        public CarOfferEntity MapToEntity(CarOfferModel model) => new()
        {
            Id = model.Id,
            Heading = model.Heading,
            ShortDescription = model.ShortDescription,
            FeaturedImageUrl = model.FeaturedImageUrl,
            UrlHandle = model.UrlHandle,
            Horsepower = model.Horsepower,
            YearOfProduction = model.YearOfProduction,
            EngineDetails = model.EngineDetails,
            DriveDetails = model.DriveDetails,
            GearboxDetails = model.GearboxDetails,
            CarDeliverylocation = model.CarDeliverylocation,
            CarReturnLocation = model.CarReturnLocation,
            PublishedDate = model.PublishedDate,
            Visible = model.Visible,
            Tarrif = _tariffMapper.MapToEntity(model.Tarrif),
            Tags = model.Tags?.Select(tag => _tagMapper.MapToEntity(tag)).ToArray() ?? Array.Empty<CarTagEntity>(),
            ImageUrls = model.ImageUrls?.Select(image => _imageUrlMapper.MapToEntity(image)).ToArray() ?? Array.Empty<ImageUrlEntity>(),
            UnavailableDates = model.UnavailableDates?.Select(period => _timePeriodMapper.MapToEntity(period)).ToArray() ?? Array.Empty<TimePeriodEntity>()
        };
    }
    
    
    
}

File: RentalApi/Infrastructure/databases/postgres/mappers/car_offer_persistence_mapper.go
package mappers

import (
	models "rental-api/Domain/models/domestic"
	"rental-api/Infrastructure/databases/postgres/entities"
	"time"

	"github.com/google/uuid"
)

type CarOfferMapper struct{}

func NewCarOfferPersistenceMapper() *CarOfferMapper {
	return &CarOfferMapper{}
}

func (m *CarOfferMapper) MapToModel(entity entities.CarOfferEntity) models.CarOfferModel {
	return models.CarOfferModel{
		Id:                 entity.ID.String(),
		Heading:            entity.Heading,
		ShortDescription:   entity.ShortDescription,
		UrlHandle:          entity.UrlHandle,
		Horsepower:         entity.Horsepower,
		YearOfProduction:   entity.YearOfProduction,
		EngineDetails:      entity.EngineDetails,
		DriveDetails:       entity.DriveDetails,
		GearboxDetails:     entity.GearboxDetails,
		PublishedDate:      entity.PublishedDate.Format(time.RFC3339),
		Visible:            entity.Visible,
		OneNormalDayPrice:  entity.OneNormalDayPrice,
		OneWeekendDayPrice: entity.OneWeekendDayPrice,
		OneWeekPrice:       entity.OneWeekPrice,
		OneMonthPrice:      entity.OneMonthPrice,
		CreatedAt:          entity.CreatedAt.Format(time.RFC3339),
		UpdatedAt:          entity.UpdatedAt.Format(time.RFC3339),
		CustodianEmail:     entity.CustodianEmail,
	}
}

func (m *CarOfferMapper) MapToEntity(model models.CarOfferModel) entities.CarOfferEntity {
	var id uuid.UUID
	if model.Id == "" {
		id = uuid.New()
	} else {
		id, _ = uuid.Parse(model.Id)
	}

	publishedDate, _ := time.Parse(time.RFC3339, model.PublishedDate)

	var createdAt time.Time
	if model.CreatedAt == "" {
		createdAt = time.Now()
	} else {
		createdAt, _ = time.Parse(time.RFC3339, model.CreatedAt)
	}

	return entities.CarOfferEntity{
		ID:                 id,
		Heading:            model.Heading,
		ShortDescription:   model.ShortDescription,
		UrlHandle:          model.UrlHandle,
		Horsepower:         model.Horsepower,
		YearOfProduction:   model.YearOfProduction,
		EngineDetails:      model.EngineDetails,
		DriveDetails:       model.DriveDetails,
		GearboxDetails:     model.GearboxDetails,
		PublishedDate:      publishedDate,
		Visible:            model.Visible,
		OneNormalDayPrice:  model.OneNormalDayPrice,
		OneWeekendDayPrice: model.OneWeekendDayPrice,
		OneWeekPrice:       model.OneWeekPrice,
		OneMonthPrice:      model.OneMonthPrice,
		CustodianId:        uuid.MustParse(model.CustodianId),
		CustodianEmail:     model.CustodianEmail,
		CreatedAt:          createdAt,
		UpdatedAt:          time.Now(),
	}
}
